### PR TITLE
Fix #2307 Add support for auto downsizing non-animated GIFs with Imag…

### PR DESCRIPTION
**Pull request by azusaaz**
The non-animated gif also got to be resized to around 250KB.

test file1// non-animated gif over 300KB-> around 250KB
single

snapcrab_noname_2018-1-5_10-56-34_no-00

test file 2 // animated gif over 300KB-> keep original
anime

snapcrab_noname_2018-1-5_10-56-21_no-00

test file 3 //  non-animated png over 500KB-> around 250KB
single

snapcrab_noname_2018-1-15_17-34-39_no-00

test file 4 // animated png(not move on IE) over 300KB-> keep original
anime

snapcrab_noname_2018-1-15_17-34-27_no-00

### CODE REVIEW

**Comment by azusaaz** [comment]
A module directory "is-animated" was under "node_modules" in my local.
However, it got a reference problem on the public ( there is no "node_modules" directory?) and was needed to be added some lines.
("define(function (require, exports, module) {  });" line and the semicolons).

So, bit messy, but I copied "is-animated" directory into "filesystem/impls/filer/lib/is-animated".
Also, it caused a result of change "+16,683 −13,523 " somehow.
Is there any way I can avoid this?

**Comment by gideonthomas** [comment]
@azusaaz, don't commit anything from node_modules into the source tree. If you aren't seeing `node_modules`, make sure you run `npm install` and your new module - `is-animated` is in your package.json. If it's missing the `define`, don't manually add it. If you only require three files from the module, copy only those three files inside an `is-animated` folder inside `src/thirdparty`. Then, change the require config here: https://github.com/mozilla/brackets/blob/master/src/main.js#L28-L59 to include a `path` to `thirdparty/is-animated/index.js` with a key of `"isAnimated"`, then add a `shim` to the same require config with a value of whatever the module exports is in the `is-animated` library. Take a look at the `"Caman"` configuration in that require config; it's very similar to that.

**Comment by azusaaz** [comment]
@gideonthomas Thank you so much for your comment! I'll fix that.

File: src/filesystem/impls/filer/FilerFileSystem.js [hunk 1]
```diff
@@ -304,14 +305,20 @@ define(function (require, exports, module) {
 
                     // If this is a big image (>250K), resize it first
                     if(Content.isResizableImage(ext) && Content.isImageTooLarge(data.length)) {
-                        ImageResizer.resize(path, data, function(err, resized) {
-                            if(err) {
-                                return callback(err);
-                            }
-
-                            data = resized;
+                        if(path.match(".*\.gif")&& isAnimated(data)) {
```
Comment (humphd) [inline src/filesystem/impls/filer/FilerFileSystem.js]: Let's move this logic (i.e., checking if this is both a `.gif` and whether it's animated or not) into `Content.isResizableImage()`, so that you don't need to do any extra `if` checks here.

File: src/filesystem/impls/filer/lib/is-animated/.npmignore [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Coverage directory used by tools like istanbul
```
Comment (humphd) [inline src/filesystem/impls/filer/lib/is-animated/.npmignore]: @gideonthomas how do you want to handle importing this code?  I'm tempted to just take the code we need, and drop the rest of these files.

Comment (gideonthomas) [reply src/filesystem/impls/filer/lib/is-animated/.npmignore]: yeah, I agree (I added a comment above), let's just take the code we need, add it to thirdparty and load it from there.

File: src/filesystem/impls/filer/FilerFileSystem.js (hunk 1 above)
Comment (azusaaz) [reply src/filesystem/impls/filer/FilerFileSystem.js]: @humphd Ohhhh. It's nice. Thank you!

Comment (humphd) [reply src/filesystem/impls/filer/FilerFileSystem.js]: One further thought, you'll either need to modify what `isResizableImage()` takes as an argument--it uses an extension now, but you'd need the entire path.  Or you could split this into two functions, and have the `isAnimated()` be a separate function that lives on `Content`.  Does that make sense?

Comment (azusaaz) [reply src/filesystem/impls/filer/FilerFileSystem.js]: @humphd I need to bring data into Content.isResizableImage() for animation check, but if I just I add "data" argument, I got a data type error. To avoid this, do I need the entire path? Is that what you mean?

It seems this one works to filter animated gif and png, but is that not nice for jpg and jpeg?
// FillerFileSystem.js line307 (added isAnimated check at the top of the original)
snapcrab_noname_2018-1-10_19-22-22_no-00

Comment (humphd) [reply src/filesystem/impls/filer/FilerFileSystem.js]: Yes, I'd pass in the whole path to a new `isAnimated` method on `Content`:

```js
if((Content.isResizableImage(ext) && !Content.isAnimated(path)) && Content.isImageTooLarge(data.length)) {
...
```

Comment (azusaaz) [reply src/filesystem/impls/filer/FilerFileSystem.js]: Oh, that's why I can have two functions...! 
I see. Thank you!

Comment (azusaaz) [reply src/filesystem/impls/filer/FilerFileSystem.js]: @humphd 
I'd like to push the latest one,
but could you let me ask about the rebase please?

++now in #2307branch
   git pull --rebase origin master
++fix conflicts
   git add *
   git rebase --continue
   git push origin #2307branch -f
   
Is that correct way?
Should I do "rebase" now, or after the whole fixing was approved?

Comment (humphd) [reply src/filesystem/impls/filer/FilerFileSystem.js]: I would do this:

```
git checkout master
git pull origin master
git checkout #2307branch
git rebase master
++fix conflicts
git add <files-you-changed-one-by-one>
git rebase --continue
git push origin #2307branch -f
```

Also, I'd recommend you rebase when you've finished everything and gotten a positive review.  Though you can do it either way.

Comment (azusaaz) [reply src/filesystem/impls/filer/FilerFileSystem.js]: @humphd Thank you so much! Then, I will do rebase later!

**Comment by azusaaz** [comment]
updated

**Comment by gideonthomas** [comment]
@azusaaz, this PR has a bunch of changes in `src/nls` that shouldn't be in here. Would you be able to fix your commits so that only the changes you've made show up?

**Comment by azusaaz** [comment]
@gideonthomas
I'm sorry about the mess. I'll fix that.

File: src/extensions/default/bramble/lib/PostMessageTransport.js [hunk 3]
```diff
@@ -107,7 +107,7 @@ define(function (require, exports, module) {
             // Not one of our custom events, re-trigger message event to LiveDev
             module.exports.trigger("message", [connId, msgObj.message]);
         } else if (msgObj.type === "connect") {
-            Browser.setListener();
+            //Browser.setListener();
```
Comment (humphd) [inline src/extensions/default/bramble/lib/PostMessageTransport.js]: Let's revert this change please, it seems unrelated to your work.

File: src/filesystem/impls/filer/lib/content.js [hunk 4]
```diff
@@ -140,7 +141,11 @@ define(function (require, exports, module) {
 
         isResizableImage: function(ext) {
             ext = FilerUtils.normalizeExtension(ext);
-            return ext === '.png' || ext === '.jpg' || ext === '.jpeg';
+            return ext === '.png' || ext === '.jpg' || ext === '.jpeg' || ext === '.gif';
+        },
+
+        isAnimated: function(data) {
```
Comment (humphd) [inline src/filesystem/impls/filer/lib/content.js]: To avoid confusion between the two uses of `isAnimated`, can you update this to:

```js
isAnimatedImage: function(data) {
    return isAnimated(data);
},
```

File: src/extensions/default/bramble/lib/PostMessageTransport.js (hunk 3 above)
Comment (azusaaz) [reply src/extensions/default/bramble/lib/PostMessageTransport.js]: When a file (any type) was uploaded or deleted, console complains about the Browser.setListnere(); line, since the function was removed.
 Is that fine to leave it as it is?
error

File: src/filesystem/impls/filer/lib/content.js (hunk 4 above)
Comment (azusaaz) [reply src/filesystem/impls/filer/lib/content.js]: Ok!

File: src/extensions/default/bramble/lib/PostMessageTransport.js (hunk 3 above)
Comment (humphd) [reply src/extensions/default/bramble/lib/PostMessageTransport.js]: Interesting, that makes no sense to me.  @gideonthomas do you remember us touching this at all?  We might have another bug here that needs to be fixed.

So to be clear: when you leave that line in (not commented out), and drag a file in, you get that error in the console?

Comment (azusaaz) [reply src/extensions/default/bramble/lib/PostMessageTransport.js]: Yes, if I keep the line, I get the error every time. Then the number of the counter is increased.
However, on "Thimble", this error doesn't happen.

A friend found that the function was removed here.
https://github.com/mozilla/brackets/commit/1049e3cbbb9762b1ac1371c0744fd6c1bead56b6

Comment (humphd) [reply src/extensions/default/bramble/lib/PostMessageTransport.js]: Thanks for letting us know about this, and researching the cause.  I've filed another bug to fix it: https://github.com/mozilla/thimble.mozilla.org/issues/2608.  Don't worry about it in your PR.

Comment (azusaaz) [reply src/extensions/default/bramble/lib/PostMessageTransport.js]: Ok! Then, I'll revert the line.

**Comment by azusaaz** [comment]
updated

File: package.json [hunk 5]
```diff
@@ -42,6 +42,7 @@
         "grunt-targethtml": "0.2.6",
         "grunt-template-jasmine-requirejs": "0.1.0",
         "grunt-usemin": "0.1.11",
+        "is-animated": "^1.1.1",
```
Comment (gideonthomas) [inline package.json]: if we're only using files we need from this in `thirdparty`, we don't need it to be included here.

File: src/config.json [hunk 6]
```diff
@@ -65,6 +65,7 @@
         "grunt-targethtml": "0.2.6",
         "grunt-template-jasmine-requirejs": "0.1.0",
         "grunt-usemin": "0.1.11",
+        "is-animated": "^1.1.1",
```
Comment (gideonthomas) [inline src/config.json]: remove this

File: src/thirdparty/is-animated/.npmignore [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Coverage directory used by tools like istanbul
```
Comment (gideonthomas) [inline src/thirdparty/is-animated/.npmignore]: remove this file

File: src/thirdparty/is-animated/.travis.yml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+language: node_js
```
Comment (gideonthomas) [inline src/thirdparty/is-animated/.travis.yml]: remove this file

File: src/thirdparty/is-animated/README.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# is-animated
```
Comment (gideonthomas) [inline src/thirdparty/is-animated/README.md]: remove this file

File: src/thirdparty/is-animated/package.json [hunk 10]
```diff
@@ -0,0 +1,101 @@
+{
```
Comment (gideonthomas) [inline src/thirdparty/is-animated/package.json]: remove this file

Comment (azusaaz) [reply src/thirdparty/is-animated/package.json]: @gideonthomas Thank you for the comments, I'll remove the part and files!

**Comment by azusaaz** [comment]
changed

**Comment by gideonthomas** [comment]
Thanks @azusaaz, this looks great! Nice work!
